src/edwards_point.rs

```diff
@@ -0,0 +1,167 @@
+use bulletproofs::r1cs::{ConstraintSystem, LinearCombination};
+use curve25519_dalek::scalar::Scalar;
+// Represents a Sonny Edwards Point using Twisted Edwards Extended Coordinates
+pub struct SonnyEdwardsPointGadget {
+    pub X: LinearCombination,
+    pub Y: LinearCombination,
+    pub Z: LinearCombination,
+    pub T: LinearCombination,
+}
+
+impl SonnyEdwardsPointGadget {
+    pub fn add(
+        self,
+        cs: &mut dyn ConstraintSystem,
+        other: SonnyEdwardsPointGadget,
+    ) -> SonnyEdwardsPointGadget {
+        // XXX: public constants should be defined at a higher level
+        let a: Scalar = Scalar::from_bytes_mod_order(zerocaf::constants::EDWARDS_A.to_bytes());
+        let d: Scalar = Scalar::from_bytes_mod_order(zerocaf::constants::EDWARDS_D.to_bytes());
+
+        // Point addition impl
+        // A = p1_x * p2_x
+        // B = p1_y * p2_y
+        // C = d*(p1_t * p2_t)
+        // D = p1_z * p2_z
+        // E = (p1_x + p1_y) * (p2_x + p2_y) + a*A + a*B
+        // F = D - C
+        // G = D + C
+        // H = B + A
+        // X3 = E * F , Y3 = G * H, Z3 = F * G, T3 = E * H
+        //
+        // Compute A
+        let (_, _, A) = cs.multiply(self.X.clone(), other.X.clone());
+
+        // Compute B
+        let (_, _, B) = cs.multiply(self.Y.clone(), other.Y.clone());
+
+        // Compute C
+        let (_, _, pt) = cs.multiply(self.T, other.T);
+        let (_, _, C) = cs.multiply(pt.into(), d.into());
+
+        // Compute D
+        let (_, _, D) = cs.multiply(self.Z, other.Z);
+
+        // Compute E
+        let E = {
+            let E1 = self.X.clone() + self.Y.clone();
+            cs.constrain(E1.clone() - self.X - self.Y);
+
+            let E2 = other.X.clone() + other.Y.clone();
+            cs.constrain(E2.clone() - other.X - other.Y);
+
+            let (_, _, E12) = cs.multiply(E1, E2);
+
+            let (_, _, aA) = cs.multiply(a.into(), A.into());
+            let (_, _, bB) = cs.multiply(a.into(), B.into());
+
+            let E = aA + bB + E12;
+            cs.constrain(E.clone() - aA - bB - E12);
+
+            E
+        };
+
+        // Compute F
+        let F = D - C;
+        cs.constrain(F.clone() - D + C);
+
+        // Compute G
+        let G = D + C;
+        cs.constrain(G.clone() - D - C);
+
+        // Compute H
+        let H = B + A;
+        cs.constrain(H.clone() - B - A);
```

Comment:
I think we can save this addition constrains just by adding a `valid_point_constrain` just before returning the value.
In that case, we ensure that the additions were done correctly if the resulting point was valid (satisfies the equation)

Reply:
I'm not sure I follow, we will need to have these constraints either way, unless you get rid of the variable assignments, but then this would make the code almost harder to read.

Reply:
What I meant is that here you check that the additions are done correctly.
I meant that we can save them if we add a constrain to check the point correctness following the Tedws formulae.

Maybe this makes no sense hahah, just thought that this will save us 2 constrains since if the resulting point is correct, it means the additions were done correctly.